datafusion/sql/tests/cases/plan_to_sql.rs

```diff
@@ -314,3 +310,78 @@ fn test_table_references_in_plan_to_sql() {
         "SELECT \"table\".id, \"table\".\"value\" FROM \"table\"",
     );
 }
+
+#[test]
+fn test_pretty_roundtrip() -> Result<()> {
+    let schema = Schema::new(vec![
+        Field::new("id", DataType::Utf8, false),
+        Field::new("age", DataType::Utf8, false),
+    ]);
+
+    let df_schema = DFSchema::try_from(schema)?;
+
+    let context = MockContextProvider::default();
+    let sql_to_rel = SqlToRel::new(&context);
+
+    let unparser = Unparser::default();
+
+    let sql_to_pretty_unparse = vec![
+        ("((id < 5) OR (age = 8))", "id < 5 OR age = 8"),
+        ("((id + 5) * (age * 8))", "(id + 5) * age * 8"),
+        ("(3 + (5 * 6) * 3)", "3 + 5 * 6 * 3"),
+        ("((3 * (5 + 6)) * 3)", "3 * (5 + 6) * 3"),
+        ("((3 AND (5 OR 6)) * 3)", "(3 AND (5 OR 6)) * 3"),
+        ("((3 + (5 + 6)) * 3)", "(3 + 5 + 6) * 3"),
+        ("((3 + (5 + 6)) + 3)", "3 + 5 + 6 + 3"),
+        ("3 + 5 + 6 + 3", "3 + 5 + 6 + 3"),
+        ("3 + (5 + (6 + 3))", "3 + 5 + 6 + 3"),
+        ("3 + ((5 + 6) + 3)", "3 + 5 + 6 + 3"),
+        ("(3 + 5) + (6 + 3)", "3 + 5 + 6 + 3"),
+        ("((3 + 5) + (6 + 3))", "3 + 5 + 6 + 3"),
+        (
+            "((id > 10) OR (age BETWEEN 10 AND 20))",
+            "id > 10 OR age BETWEEN 10 AND 20",
+        ),
+        (
+            "((id > 10) * (age BETWEEN 10 AND 20))",
+            "(id > 10) * (age BETWEEN 10 AND 20)",
+        ),
+        ("id - (age - 8)", "id - (age - 8)"),
+        ("((id - age) - 8)", "id - age - 8"),
+        ("(id OR (age - 8))", "id OR age - 8"),
+        ("(id / (age - 8))", "id / (age - 8)"),
+        ("((id / age) * 8)", "id / age * 8"),
+        ("((age + 10) < 20) IS TRUE", "(age + 10 < 20) IS TRUE"),
+        (
+            "(20 > (age + 5)) IS NOT FALSE",
+            "(20 > age + 5) IS NOT FALSE",
+        ),
+        ("(true AND false) IS FALSE", "(true AND false) IS FALSE"),
+        ("true AND (false IS FALSE)", "true AND false IS FALSE"),
+    ];
+
+    for (sql, pretty) in sql_to_pretty_unparse.iter() {
+        let sql_expr = Parser::new(&GenericDialect {})
+            .try_with_sql(sql)?
+            .parse_expr()?;
+        let expr =
+            sql_to_rel.sql_to_expr(sql_expr, &df_schema, &mut PlannerContext::new())?;
+        let round_trip_sql = unparser.expr_to_sql(&expr)?.to_string();
+        assert_eq!(pretty.to_string(), round_trip_sql);
+
+        // verify that the pretty string parses to the same underlying Expr
+        let pretty_sql_expr = Parser::new(&GenericDialect {})
+            .try_with_sql(pretty)?
+            .parse_expr()?;
+
+        let pretty_expr = sql_to_rel.sql_to_expr(
+            pretty_sql_expr,
+            &df_schema,
+            &mut PlannerContext::new(),
+        )?;
+
+        assert_eq!(expr.to_string(), pretty_expr.to_string());
```

Comment:
Shouldn't this me veriying that the expr are the same (not just that the string is the same)?

Reply:
But they can't be the same, the whole idea behind removing parenthesis is to maintain the final result regardless of which operations get executed first/has the parenthesis aka associativity.

Here's an example from the tests:
` ("((id + 5) * (age * 8))", "(id + 5) * age * 8"),`

Here's the tree of the first/original expr:
```
	    *
	   / \
	  +   * 
	/  \  / \
      id  5 age 8
```

And here's the cleaned one
```
	       *
	      / \
	     *   8
	   / \
	  +   age
	 / \
	id  5
```
These obviously don't match, but it doesn't matter because multiplication is associative. Let (id+5) be x -> (x * age) * 8 is the exact same as x * (age * 8). Almost all operators are associative except for division and subtraction, which are handled.

Comparing the string representation seemed like the best solution.

Reply:
If you want to maintain logical equivalence then the only case we remove parenthesis would be that the parenthesis give priority to left operations which is the natural associativity of all binary operators (except `=` I think)*. Something like `(a + b) + c` for example

\* https://www.postgresql.org/docs/7.2/sql-precedence.html

Reply:
Thank you for the response @MohamedAbdeen21  -- this makes sense

🤔  I am not sure if maintaining the structural equivalence of the input/output is impotant or if just the semantics are important to perserve

Perhaps @goldmedal  @seddonm1  or @phillipleblanc or @devinjdangelo  has thoughts on what the preferred behavior is

Reply:
I think it's related to https://github.com/apache/datafusion/pull/11186#discussion_r1664797700. If our main purpose is providing SQL for full pushdown to other data sources, this behavior makes sense to me. For my current project (Wren), it's acceptable.

However, if our purpose is human readability, I think preserving the semantics is important. Maybe we can just remove the outermost brackets (they're exactly redundant) for a calculation expression like:

```
((3 + (5 + 6)) + 3) -> (3 + (5 + 6)) + 3
```

I'm not sure, but I think it might be more similar to what a human would write.

Reply:
I can see the pros and cons of each approach and to be honest that's part of why I wanted to keep the original `expr_to_sql` and add a new method instead.

I believe we can keep the parse tree structure if that's more important (haven't tried that yet); but again, as you've noted, that may not be very human-friendly. So it's up to the people who use the unparser the most to decide.

Reply:
I don't think maintaining structural equivalence is important as long as the correct answer would still be generated if the SQL were run.

But maybe this is exposing two opposing(?) goals of the `unparser` - the goal for using it to generate SQL that will be run in another SQL query engine. And another goal for displaying it for human readability/debuggability. I'm in the first camp - it doesn't matter to me what the resulting SQL looks like as long as it runs correctly (although having nicer SQL does make debugging easier).

However, I would expect that we could serve both goals with a single `unparser` implementation. Any work that is done to improve the SQL for human readability shouldn't come at the expense of making it invalid SQL for running in a separate query engine. That is why I thought it would be better to have just a single `expr_to_sql` function instead of splitting the functionality.

Reply:
Since the primary goal is pushdown for other systems, we have two options.

1. Completely ignore this issue. Systems can have very different precedence rules, and therefore, we can not guarantee correctness once some parentheses are removed. Also, it can be hard to decide which parentheses are "safe" to remove.

2. Informing DF of your target system operator precedence through implementing a dialect. This can be a lot of work for the dev, but I can see this being a desired option for some people; as Phillip pointed out, It can make debugging easier.

Reply:
> But maybe this is exposing two opposing(?) goals of the unparser - the goal for using it to generate SQL that will be run in another SQL query engine. And another goal for displaying it for human readability/debuggability. 

I think this is an excellent summary of the issue and given these two different goals there would be two different ideal outcomes. 

Perhaps as @MohamedAbdeen21  suggests we can add some sort of API to allow the user to decide. We could potentially use `Unparser` for this (though I don't think `Unparser` is part of the public API yet). 

https://github.com/apache/datafusion/blob/45599ce310aa6270813091a5c3288abcd7541f59/datafusion/sql/src/unparser/mod.rs#L30-L32

So something like
```rust
// default unparser generates fully parenthesized sql:
let unparser = Unparser::default();


// can also create an unparser that generates pretty sql
let unparser = Unparser::default()
 // enable pretty-printing of the sql easier suited to human consumption
 .with_pretty(true);

Reply:
That makes sense to me. I didn't think about how different query engines might have different precedence rules, so having this functionality split now makes a lot more sense.

Making it explicit: Removing parenthesis according to the precedence rules of DataFusion might make it invalid SQL for other query engines with different precedence rules, even if its valid for DataFusion. That would break one of the use-cases for the unparser, so allowing people to opt-in to that behavior is desirable.

The proposed API by @alamb looks good to me.

Reply:
pushed the proposed API, updated docs, examples and tests